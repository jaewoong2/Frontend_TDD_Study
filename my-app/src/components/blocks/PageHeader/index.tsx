import { TopNav } from 'components/atoms';
import { useEffect, useState } from 'react';
import { Link, useLocation } from 'react-router-dom';

function PageHeader() {
  const [pathName, setPathName] = useState('');
  const { pathname } = useLocation();

  useEffect(() => {
    setPathName(() => {
      switch (pathname) {
        case '/':
          return '할 일 목록';
        case '/add':
          return '할 일 추가';
        default: {
          if (pathname.startsWith('/detail')) {
            return '할 일 상세';
          }
          return 'Error!';
        }
      }
    });
  }, [pathname]);
  return <TopNav left={<Link to="/">Back</Link>}>{pathName}</TopNav>;
}

export default PageHeader;
