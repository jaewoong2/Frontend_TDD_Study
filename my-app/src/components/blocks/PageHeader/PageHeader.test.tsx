import { fireEvent, render, screen } from '@testing-library/react';
import 'jest-styled-components';
import userEvent from '@testing-library/user-event';
import { BrowserRouter } from 'react-router-dom';
import PageHeader from './index';

describe('<PageHeader />', () => {
  it('Renders Component Correctly', () => {
    const { container } = renderWithRouter(<PageHeader />, { route: '/' });

    const label = screen.getByText('할 일 목록');
    expect(label).toBeInTheDocument();
    const goBack = screen.getByText('Back');
    expect(goBack).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('Renders Component Correctly with /add Url', () => {
    renderWithRouter(<PageHeader />, { route: '/add' });

    const label = screen.getByText('할 일 추가');
    expect(label).toBeInTheDocument();
    const goBack = screen.getByText('Back');
    expect(goBack).toBeInTheDocument();
  });

  it('Renders Component Correctly with /detail Url', () => {
    renderWithRouter(<PageHeader />, { route: '/detail/1' });

    const label = screen.getByText('할 일 상세');
    expect(label).toBeInTheDocument();
    const goBack = screen.getByText('Back');
    expect(goBack).toBeInTheDocument();
  });
  it('Renders Component Correctly with /add Url', () => {
    renderWithRouter(<PageHeader />, { route: '/add' });

    const label = screen.getByText('할 일 추가');
    expect(label).toBeInTheDocument();
    const goBack = screen.getByText('Back');
    expect(goBack).toBeInTheDocument();
  });

  it('Renders Component Correctly with /Error Url', () => {
    renderWithRouter(<PageHeader />, { route: '/error' });

    const label = screen.getByText('Error!');
    expect(label).toBeInTheDocument();
    const goBack = screen.getByText('Back');
    expect(goBack).toBeInTheDocument();
  });

  it('Renders Component Correctly with Goback Button ', async () => {
    renderWithRouter(<PageHeader />, { route: '/error' });

    const errorLabel = screen.getByText('Error!');
    expect(errorLabel).toBeInTheDocument();

    const goBack = screen.getByText('Back') as HTMLAnchorElement;
    expect(goBack).toBeInTheDocument();
    fireEvent.click(goBack);
    const label = screen.getByText('할 일 목록');
    expect(label).toBeInTheDocument();
  });
});

// needs for react-router-dom v6
const renderWithRouter = (ui: React.ReactElement, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);

  return {
    user: userEvent.setup(),
    ...render(ui, { wrapper: BrowserRouter }),
  };
};
