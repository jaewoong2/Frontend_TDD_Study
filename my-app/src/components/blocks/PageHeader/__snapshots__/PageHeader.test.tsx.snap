// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageHeader /> Renders Component Correctly 1`] = `
.c0 {
  width: 100%;
  height: 40px;
  background-color: #9bceee;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-radius: 8px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .box {
  width: 100%;
}

.c0 .center {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div>
  <nav
    class="c0"
  >
    <div
      class="box left"
    >
      <a
        href="/"
      >
        Back
      </a>
    </div>
    <div
      class="box center"
    >
      할 일 목록
    </div>
    <div
      class="box right"
    />
  </nav>
</div>
`;
